File: ProductKey/NET/source.cs
using System;
using Microsoft.Win32;
using System.Reflection;
using System.Collections;

[assembly: AssemblyVersion("2.0.0.0")]

namespace ProductKey {
  internal sealed class Program {
    internal static String DecodeProductKey(Byte[] digitalProductId) {
      Char[] map = ("BCDFGHJKMPQRTVWXY2346789").ToCharArray();
      Char[] key = new Char[29];
      ArrayList raw = new ArrayList();
      
      for (Int32 i = 52; i < 67; i++)
        raw.Add(digitalProductId[i]);
        
      for (Int32 i = 28; i >= 0; i--) {
        if ((i + 1) % 6 == 0) key[i] = '-';
        else {
          Int32 k = 0;
          for (Int32 j = 14; j >= 0; j--) {
            k = (k * 256) ^ (Byte)raw[j];
            raw[j] = (Byte)(k / 24);
            k %= 24;
            key[i] = map[k];
          }
        }
      }
      return new String(key);
    }
    
    internal static void Main() {
      using (RegistryKey rk =
        Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Windows NT\CurrentVersion")) {
        Byte[] val = rk.GetValue("DigitalProductId") as Byte[];
        Console.WriteLine(DecodeProductKey(val));
      }
    }
  }
}

File: ProductKey/NET/source.vb
Imports System
Imports Microsoft.Win32
Imports System.Reflection
Imports System.Collections

<assembly: AssemblyVersion("2.0.0.0")>

Namespace ProductKey
  Class Program
    Friend Shared Function DecodeProductKey(ByVal digitalProductId As Byte()) As String
      Dim map As Char() = ("BCDFGHJKMPQRTVWXY2346789").ToCharArray()
      Dim key As Char() = New Char(28) {}
      Dim raw As New ArrayList()
      
      For i As Int32 = 52 To 66
        raw.Add(digitalProductId(i))
      Next
      
      For i As Int32 = 28 To 0 Step - 1
        If (i + 1) Mod 6 = 0 Then
          key(i) = "-"C
        Else
          Dim k As Int32 = 0
          For j As Int32 = 14 To 0 Step - 1
            k = (k * 256) Xor CByte(raw(j))
            raw(j) = CByte(k \ 24)
            k = k Mod 24
            key(i) = map(k)
          Next
        End If
      Next
      Return New String(key)
    End Function
    
    Shared Sub Main()
      Using rk As RegistryKey = _
        Registry.LocalMachine.OpenSubKey("SOFTWARE\Microsoft\Windows NT\CurrentVersion")
        Dim val As Byte() = TryCast(rk.GetValue("DigitalProductId"), Byte())
        Console.WriteLine(DecodeProductKey(val))
      End Using
    End Sub
  End Class
End Namespace
